tests/e2e/improved_csi_idempotency.go

```diff
@@ -0,0 +1,572 @@
+	}()
+
+	ginkgo.By("Creating PVCs using the Storage Class")
+	framework.Logf("VOLUME_OPS_SCALE is set to %v", volumeOpsScale)
+	for i := 0; i < volumeOpsScale; i++ {
+		fmt.Printf("Creating %v pvc ", i)
+		accessMode := v1.ReadWriteOnce
+
+		// Check if it is file volumes setups
+		if rwxAccessMode {
+			accessMode = v1.ReadWriteMany
+		}
+		pvclaims[i], err = createPVC(client, namespace, nil, "", storageclass, accessMode)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+	}
+
+	if serviceName == "CSI" {
+		ginkgo.By("Stopping CSI driver")
+		// TODO: Stop printing csi logs on the console
+		// collectPodLogs(ctx, client, csiSystemNamespace)
+		err = updateDeploymentReplicawithWait(client, 0, vSphereCSIControllerPodNamePrefix, csiSystemNamespace)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		isServiceStopped = true
+
+		defer func() {
+			if isServiceStopped {
+				framework.Logf("Starting CSI driver")
+				err = updateDeploymentReplicawithWait(client, 1, vSphereCSIControllerPodNamePrefix, csiSystemNamespace)
+				gomega.Expect(err).NotTo(gomega.HaveOccurred())
+				// Wait for the CSI Pods to be up and Running
+				list_of_pods, err := fpod.GetPodsInNamespace(client, csiSystemNamespace, ignoreLabels)
+				gomega.Expect(err).NotTo(gomega.HaveOccurred())
+				num_csi_pods := len(list_of_pods)
+				fpod.WaitForPodsRunningReady(client, csiSystemNamespace, int32(num_csi_pods), 0, pollTimeout, ignoreLabels)
+			}
+		}()
+		framework.Logf("Starting CSI driver")
+		err = updateDeploymentReplicawithWait(client, 1, vSphereCSIControllerPodNamePrefix, csiSystemNamespace)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		// Wait for the CSI Pods to be up and Running
+		list_of_pods, err := fpod.GetPodsInNamespace(client, csiSystemNamespace, ignoreLabels)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		num_csi_pods := len(list_of_pods)
+		fpod.WaitForPodsRunningReady(client, csiSystemNamespace, int32(num_csi_pods), 0, pollTimeout, ignoreLabels)
+		isServiceStopped = false
+
+		ginkgo.By(fmt.Sprintf("Sleeping for %v seconds to allow full sync finish", fullSyncWaitTime))
+		time.Sleep(time.Duration(fullSyncWaitTime) * time.Second)
+	} else {
+		ginkgo.By(fmt.Sprintf("Stopping %v on the vCenter host", serviceName))
+		vcAddress := e2eVSphere.Config.Global.VCenterHostname + ":" + sshdPort
+		err = invokeVCenterServiceControl(stopOperation, serviceName, vcAddress)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		isServiceStopped = true
+		err = waitVCenterServiceToBeInState(serviceName, vcAddress, stopMessage)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+		defer func() {
+			if isServiceStopped {
+				ginkgo.By(fmt.Sprintf("Starting %v on the vCenter host", serviceName))
+				err = invokeVCenterServiceControl(startOperation, serviceName, vcAddress)
+				gomega.Expect(err).NotTo(gomega.HaveOccurred())
+				err = waitVCenterServiceToBeInState(serviceName, vcAddress, startMessage)
+				gomega.Expect(err).NotTo(gomega.HaveOccurred())
```

Comment:
after this you should set `isServiceStopped` to `false`, other wise this block in `BeforeEach()` will try to start it again

Reply:
Added